repository/src/lib.rs

```diff
@@ -115,39 +110,134 @@ pub trait DistributedRepository<T: RawRepository>: Send + Sync {
+            .locate_branch(&FINALIZED_BRANCH_NAME.into())
+            .await?;
+
+        // Check if the `work` branch is rebased on top of the `finalized` branch.
+        if self
+            .raw
+            .find_merge_base(&last_header_commit, &work_commit)
+            .await?
+            != last_header_commit
+        {
+            return Err(anyhow!(
+                "branch {} should be rebased on {}",
+                WORK_BRANCH_NAME,
+                FINALIZED_BRANCH_NAME
+            ));
+        }
+
+        // Fetch and convert commits
+        let commits = self.raw.list_ancestors(&work_commit, Some(256)).await?;
+        let position = commits
+            .iter()
+            .position(|c| *c == last_header_commit)
+            .expect("TODO: handle the case where it exceeds the limit.");
+
+        // commits starting from the very next one to the last finalized block.
+        let commits = stream::iter(commits.iter().take(position).rev().cloned().map(|c| {
+            let raw = &self.raw;
+            async move { raw.read_semantic_commit(&c).await.map(|x| (x, c)) }
+        }))
+        .buffered(256)
+        .collect::<Vec<_>>()
+        .await;
+        let commits = commits.into_iter().collect::<Result<Vec<_>, _>>()?;
+        let commits = commits
+            .into_iter()
+            .map(|(commit, hash)| {
+                from_semantic_commit(commit, &last_header)
+                    .map_err(|e| (e, hash))
+                    .map(|x| (x, hash))
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|(error, hash)| anyhow!("failed to convert the commit {}: {}", hash, error))?;
+
+        // Check the validity of the commit sequence
+        let mut verifier = CommitSequenceVerifier::new(last_header.clone())
+            .map_err(|e| anyhow!("verification error on commit {}: {}", last_header_commit, e))?;
+        for (commit, hash) in commits.iter() {
+            verifier
+                .apply_commit(commit)
+                .map_err(|e| anyhow!("verification error on commit {}: {}", hash, e))?;
+        }
+
+        // Check whether the commit sequence is in the transaction phase.
+        let mut transactions = Vec::new();
+
+        for (commit, _) in commits {
+            if let Commit::Transaction(t) = commit {
+                transactions.push(t.clone());
+            } else {
+                return Err(anyhow!(
+                    "branch {} is not in the transaction phase",
+                    WORK_BRANCH_NAME
+                ));
+            }
+        }
+
+        let agenda_commit = Commit::Agenda(Agenda {
+            author,
+            timestamp: get_timestamp(),
+            hash: Agenda::calculate_hash(last_header.height + 1, &transactions),
+        });
+        let semantic_commit = to_semantic_commit(&agenda_commit, &last_header);
+
+        self.raw.checkout_clean().await?;
+        self.raw.checkout(&FINALIZED_BRANCH_NAME.into()).await?;
```

Comment:
Why do you checkout to `main` just before creating an agenda commit?

Reply:
fixed